cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java

```diff
@@ -318,35 +346,57 @@ private void processMetrics() {
           topicMessageIds.put(topicIdMetaKey, currentMessageId);
         }
 
-        // Skip persisting if the records is empty
-        if (records.isEmpty()) {
-          return;
+        // Try to persist metrics and messageId's of the last metrics to be persisted if no other thread is persisting
+        tryPersist();
+
+        long endTime = System.currentTimeMillis();
+        if (endTime - lastMetricTimeMillis > maxDelayMillis) {
+          // Don't sleep if falling behind
+          return 0L;
+        } else {
+          long timeSpent = endTime - startTime;
+          return Math.max(0L, metricsProcessIntervalMillis - timeSpent);
         }
+      } catch (Exception e) {
+        LOG.warn("Failed to process metrics. Will be retried in next iteration.", e);
+      }
+      return metricsProcessIntervalMillis;
+    }
 
-        // Ensure there's only one thread can persist records and messageId's when the persistingFlag is false.
-        // The thread set persistingFlag to true when it starts to persist.
-        if (!persistingFlag.compareAndSet(false, true)) {
-          LOG.trace("Cannot persist because persistingFlag is taken by another thread.");
-          return;
+    /**
+     * Persist metrics and messageId's of the last metrics to be persisted if no other thread is persisting
+     */
+    private void tryPersist() {
+      // Ensure there's only one thread can persist metricsFromAllTopics and messageId's.
+      // If persistingFlag is false, set it to true and start persisting. Otherwise, log and return.
+      if (!persistingFlag.compareAndSet(false, true)) {
+        LOG.trace("Cannot persist because persistingFlag is taken by another thread.");
```

Comment:
This message is a bit misleading. It is not cannot persist. It is this thread is not the persisting thread, hence letting other threads to do the persisting of metrics.